File: android/app/src/main/java/br/com/querosermb/presentation/exchangedetail/ExchangeDetailViewModel.kt
package br.com.querosermb.presentation.exchangedetail

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import br.com.querosermb.core.cache.ExchangeDetailCaching
import br.com.querosermb.domain.model.Currency
import br.com.querosermb.domain.model.Exchange
import br.com.querosermb.domain.usecase.GetExchangeAssetsUseCase
import br.com.querosermb.domain.usecase.GetExchangeDetailUseCase
import br.com.querosermb.presentation.ViewState
import br.com.querosermb.presentation.utils.toUiText
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ExchangeDetailViewModel @Inject constructor(
    private val getExchangeDetail: GetExchangeDetailUseCase,
    private val getExchangeAssets: GetExchangeAssetsUseCase,
    private val detailCache: ExchangeDetailCaching
) : ViewModel() {

    companion object {
        private const val CACHE_TTL_MS = 90_000L
    }

    private val _detailState = MutableStateFlow<ViewState<Exchange>>(ViewState.Loading)
    val detailState: StateFlow<ViewState<Exchange>> = _detailState.asStateFlow()

    private val _assetsState = MutableStateFlow<ViewState<List<Currency>>>(ViewState.Loading)
    val assetsState: StateFlow<ViewState<List<Currency>>> = _assetsState.asStateFlow()

    private var loadJob: Job? = null

    fun load(id: Int) {
        if (loadJob?.isActive == true) return
        loadJob = viewModelScope.launch {
            doLoad(id)
        }
    }

    fun triggerLoad(id: Int) {
        loadJob?.cancel()
        loadJob = viewModelScope.launch {
            doLoad(id)
        }
    }

    private suspend fun doLoad(id: Int) {
        val cached = detailCache.get(id, CACHE_TTL_MS)
        if (cached != null) {
            _detailState.value = ViewState.Success(cached.detail)
            _assetsState.value = if (cached.assets.isEmpty()) {
                ViewState.Empty
            } else {
                ViewState.Success(cached.assets)
            }
            return
        }

        _detailState.value = ViewState.Loading
        _assetsState.value = ViewState.Loading

        val (detailResult, assetsResult) = coroutineScope {
            val detailDeferred = async { runCatching { getExchangeDetail.execute(id) } }
            val assetsDeferred = async { runCatching { getExchangeAssets.execute(id) } }
            detailDeferred.await() to assetsDeferred.await()
        }

        val detail = detailResult.getOrNull()
        val assets = assetsResult.getOrNull()

        if (detail != null && assets != null) {
            detailCache.set(id, detail, assets)
        }

        _detailState.value = detailResult.fold(
            onSuccess = { ViewState.Success(it) },
            onFailure = { ViewState.Error(it.toUiText()) }
        )

        _assetsState.value = assetsResult.fold(
            onSuccess = { if (it.isEmpty()) ViewState.Empty else ViewState.Success(it) },
            onFailure = { ViewState.Error(it.toUiText()) }
        )
    }
}

File: ios/QuerosermB/Presentation/ExchangeDetail/ExchangeDetailViewModel.swift
import Foundation

// MARK: - ExchangeDetail ViewModel

@MainActor
final class ExchangeDetailViewModel {
    @Published var detailState: ViewState<Exchange> = .loading
    @Published var assetsState: ViewState<[Currency]> = .loading

    private let getExchangeDetail: GetExchangeDetailUseCase
    private let getExchangeAssets: GetExchangeAssetsUseCase
    private let detailCache: ExchangeDetailCaching

    /// TTL sugerido no plano: 60–120 s.
    private static let cacheTTL: TimeInterval = 90

    private var currentLoadTask: Task<Void, Never>?

    init(
        getExchangeDetail: GetExchangeDetailUseCase,
        getExchangeAssets: GetExchangeAssetsUseCase,
        detailCache: ExchangeDetailCaching
    ) {
        self.getExchangeDetail = getExchangeDetail
        self.getExchangeAssets = getExchangeAssets
        self.detailCache = detailCache
    }

    /// Dispara o carregamento cancelando qualquer load anterior em andamento.
    /// Use nos botões de retry — o `.task {}` da view usa `load(exchange:)` diretamente.
    func triggerLoad(exchange: Exchange) {
        currentLoadTask?.cancel()
        currentLoadTask = Task {
            await load(exchange: exchange)
        }
    }

    func load(exchange: Exchange) async {
        let id = exchange.id
        if let cached = detailCache.get(exchangeId: id, ttl: Self.cacheTTL) {
            detailState = .success(cached.detail)
            assetsState = cached.assets.isEmpty ? .empty : .success(cached.assets)
            return
        }

        detailState = .loading
        assetsState = .loading

        async let detailResult = fetchDetail(id: id)
        async let assetsResult = fetchAssets(id: id)
        let (detailRes, assetsRes) = await (detailResult, assetsResult)

        switch (detailRes, assetsRes) {
        case (.success(let detail), .success(let assets)):
            detailCache.set(exchangeId: id, detail: detail, assets: assets)
            detailState = .success(detail)
            assetsState = assets.isEmpty ? .empty : .success(assets)
        case (.failure(let err), .success(let assets)):
            detailState = .error(Self.message(for: err, fallback: Strings.Error.detail))
            assetsState = assets.isEmpty ? .empty : .success(assets)
        case (.success(let detail), .failure(let err)):
            detailState = .success(detail)
            assetsState = .error(Self.message(for: err, fallback: Strings.Error.assets))
        case (.failure(let dErr), .failure(let aErr)):
            detailState = .error(Self.message(for: dErr, fallback: Strings.Error.detail))
            assetsState = .error(Self.message(for: aErr, fallback: Strings.Error.assets))
        }
    }

    private func fetchDetail(id: Int) async -> Result<Exchange, Error> {
        do {
            return .success(try await getExchangeDetail.execute(id: id))
        } catch {
            return .failure(error)
        }
    }

    private func fetchAssets(id: Int) async -> Result<[Currency], Error> {
        do {
            return .success(try await getExchangeAssets.execute(id: id))
        } catch {
            return .failure(error)
        }
    }

    private static func message(for error: Error, fallback: String) -> String {
        if let ne = error as? NetworkError {
            return ne.errorDescription ?? fallback
        }
        return error.localizedDescription
    }
}
